src/explore-education-statistics-common/src/modules/table-tool/utils/applyTableHeadersOrder.ts

```diff
@@ -0,0 +1,439 @@
+          updatedGroups,
+        )
+      : getAdditionalGroups(
+          originalGroupsOnOtherAxis,
+          originalGroups,
+          updatedGroups,
+        );
+
+  // If the groups were reordered append new groups to the end, else maintain the default order.
+  return groupsAreReordered || originalGroupsWereReordered
+    ? [...orderedGroups, ...additionalGroups]
+    : [...orderedGroups, ...additionalGroups].sort((a, b) => {
+        return (
+          updatedGroupsOrder.indexOf(getFilterType(a[0])) -
+          updatedGroupsOrder.indexOf(getFilterType(b[0]))
+        );
+      });
+};
+
+/**
+ * Apply previous ordering, from the saved order and reordering, to updated table headers.
+ */
+const applyTableHeadersOrder = ({
+  originalTableHeaders,
+  reorderedTableHeaders,
+  updatedTableHeaders,
+}: {
+  originalTableHeaders?: TableHeadersConfig;
+  reorderedTableHeaders?: TableHeadersConfig;
+  updatedTableHeaders: TableHeadersConfig;
+}): TableHeadersConfig => {
+  if (!reorderedTableHeaders && !originalTableHeaders) {
+    return updatedTableHeaders;
+  }
+
+  const allOrderedColumnGroups = orderGroups({
+    axis: 'column',
+    originalTableHeaders,
+    reorderedTableHeaders,
+    updatedTableHeaders,
+  });
+
+  const allOrderedRowGroups = orderGroups({
+    axis: 'row',
+    originalTableHeaders,
+    reorderedTableHeaders,
+    updatedTableHeaders,
+  });
+
+  const allOrderedColumnGroupsLength = allOrderedColumnGroups.length;
+  const allOrderedRowGroupsLength = allOrderedRowGroups.length;
+
+  // For safety - can't think how this would be possible, but just in case, return the updated headers so the table doesn't break.
+  if (!allOrderedColumnGroupsLength && !allOrderedRowGroupsLength) {
+    return updatedTableHeaders;
+  }
+
+  // Make sure columns is populated by taking the first of the rows groups
+  if (!allOrderedColumnGroupsLength) {
+    return {
+      columnGroups: [],
+      columns: allOrderedRowGroups[0],
+      rowGroups: allOrderedRowGroups.filter(
+        (_, index) => index > 0 && index < allOrderedRowGroupsLength - 1,
+      ),
+      rows: allOrderedRowGroups[allOrderedRowGroupsLength - 1],
+    };
+  }
+
+  // Make sure rows is populated by taking the first of the columns groups
+  if (!allOrderedRowGroupsLength) {
+    return {
+      columnGroups: allOrderedColumnGroups.filter(
+        (_, index) => index > 0 && index < allOrderedColumnGroupsLength - 1,
+      ),
+      columns: allOrderedColumnGroups[allOrderedColumnGroupsLength - 1],
+      rowGroups: [],
+      rows: allOrderedColumnGroups[0],
+    };
+  }
+
+  return {
+    columnGroups: allOrderedColumnGroups.filter(
+      (_, index) => index !== allOrderedColumnGroupsLength - 1,
+    ),
+    columns: allOrderedColumnGroups[allOrderedColumnGroupsLength - 1],
+    rowGroups: allOrderedRowGroups.filter(
+      (_, index) => index < allOrderedRowGroupsLength - 1,
+    ),
+    rows: allOrderedRowGroups[allOrderedRowGroupsLength - 1],
+  };
```

Comment:
Could use similar refactorings with `slice` and `last` to clean a lot of this up

Reply:
Changed to use slice but not last for the reason above.